last_task/server/server.go

```diff
@@ -0,0 +1,206 @@
+package main
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+
+	_ "github.com/lib/pq"
+)
+
+type DB struct {
+	Conn *sql.DB
+}
+
+const (
+	id_key      = "3b2ea0ea927a17e540f2330bcafdaba3" //my id_key for api.openweathermap.org
+	serverPort  = 8080
+	dbType      = "postgres"
+	dbPort      = 5432
+	table       = "requests"
+	userDB      = "root"
+	pwdDB       = "root"
+	hostDB      = "database"
+	weatherHost = "https://api.openweathermap.org/data/2.5/weather?q=%s&appid=%s" //with get
+)
+
+type Request struct {
+	ID          int    `db:"id"`
+	City        string `db:"city"`
+	RequestTime string `db:"time"`
+	Temperature string `db:"temperature"`
+}
+
+func connString(dbType, user, pwd, dbHost string, port int, dbName string) string {
+	return fmt.Sprintf("%s://%s:%s@%s:%d/%s?sslmode=disable", dbType, user, pwd, dbHost, port, dbName)
+}
+
+func rowsToRequest(rows *sql.Rows) (request []Request, err error) {
+	Requests := make([]Request, 0)
+	for rows.Next() {
+		u := &Request{}
+		if err = rows.Scan(&u.ID, &u.City, &u.RequestTime, &u.Temperature); err != nil {
+			return nil, errors.Wrap(err, "failed to list all requests (scan)")
+		}
+		Requests = append(Requests, *u)
+	}
+	return Requests, nil
+}
+
+func (db *DB) ListAll() ([]Request, error) {
+	q := "select * from requests"
+	rows, err := db.Conn.Query(q)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list all requests")
+	}
+	return rowsToRequest(rows)
+}
+func (db *DB) AddRequest(cityName string, time string, temperature string) (int64, error) {
+	q := "INSERT INTO requests (city, time, temperature) VALUES ($1, $2, $3) RETURNING id;"
+	var insertedId int64
+	err := db.Conn.QueryRow(q, cityName, time, temperature).Scan(&insertedId)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to add request")
+	}
+
+	return insertedId, nil
+}
+
+func (db *DB) UpdateRequestTemp(id int64, temperature string) error {
+	q := "UPDATE  requests SET  temperature=$1 WHERE id=$2 ;"
+	if _, err := db.Conn.Exec(q, temperature, id); err != nil {
+		return errors.Wrap(err, "failed to update request")
+	}
+	return nil
+}
+
+func ConnToDB() (*DB, error) {
+	connStr := connString(dbType, userDB, pwdDB, hostDB, dbPort, table)
+	c, err := sql.Open(dbType, connStr)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to connect to DB")
+	}
+	db := &DB{
+		Conn: c,
+	}
+	return db, nil
+
+}
+func weatherFromOpenweathermap(w http.ResponseWriter, cityName string) ([]byte, error) {
```

Comment:
Also, better to split logic by files or even packages: one pkg for database, another for API requests and so on